{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b6507bdb-3d0f-45de-a47a-8d1f3717e76a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import allel\n",
    "import moments.LD\n",
    "import scipy.interpolate\n",
    "import msprime as ms\n",
    "import matplotlib as plt\n",
    "from Windowing import *\n",
    "from Locus_selection import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a238ae4-ae59-46a7-9097-ac7b28866b37",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Preprocessing of 1000G VCF files\n",
    "1. Apply strict mask\n",
    "    - Some sites are not reliably mapped in hg38 and need to be excluded.\n",
    "    - These regions include: repeat regions, pseudogenes, centromeric, and telomeric regions.\n",
    "2. Select only biallelic SNP sites\n",
    "    - A VCF can contain mult-allelic SNPs and structure variants (indels, tandem repeats, etc)\n",
    "    - The biallelic SNP sites are the easiest to work with, and most studies focus on them. \n",
    "\n",
    "\n",
    "\n",
    "This process can take quite a long time, I saved the processed VCFs under ~/projects/ctb-sgravel/data/30x1000G_biallelic_strict_masked <br>\n",
    "The example preprocessing script is provided below and also in the directory's README.md."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fabc073e-9fcf-454f-b7fa-beae6df4e553",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sh\n",
    "### in bash, preprocess using bcftools\n",
    "### replace variable path\n",
    "chr22_vcf_path='/home/alouette/projects/ctb-sgravel/data/1000G/20201028_CCDG_14151_B01_GRM_WGS_2020-08-05_chr22.recalibrated_variants.vcf.gz'\n",
    "strict_mask_1000G='/home/alouette/projects/ctb-sgravel/data/1000G/masks_and_annotations/20160622.allChr.mask.bed'\n",
    "IntegerT=6\n",
    "module load bcftools\n",
    "bcftools --version | head -1\n",
    "### -Oz output compressed VCF format\n",
    "### -m2 -M2 -v to only view biallellic SNPs\n",
    "### -R subset VCF from region bed file\n",
    "### --threads parallelization currently only works for compression\n",
    "bcftools view -i 'FILTER==\"PASS\"' -Oz -m2 -M2 -v snps -R ${strict_mask_1000G} ${chr22_vcf_path} --threads ${IntegerT} -o \"chr22.strict_masked.vcf.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dcbc011-e194-4fcd-bdc7-9927a96b14c9",
   "metadata": {},
   "source": [
    "bcftools manual: http://samtools.github.io/bcftools/bcftools.html#expressions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48adfd5e-cff0-4e8f-a1c4-aae77d6211c0",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Optional: covert the VCF to zarr format for faster retrieval in scikit-allele\n",
    "In sgkit, there might be better approaches. <br>\n",
    "I also completed this step and stored the output in ~/projects/ctb-sgravel/data/30x1000G_biallelic_strict_masked/zarrFormat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95655f44-4122-4ce3-9002-624dd19d5e6b",
   "metadata": {},
   "source": [
    "### Population information retrieval\n",
    "1000G stored its population information in a txt file: ~/projects/ctb-sgravel/data/1000G/population_info/20130606_g1k_3202_samples_ped_population.txt <br>\n",
    "This file includes family ID, sample ID, parent ID, sex, population, superpopulation. <br>\n",
    "There are some family trio in 1000G data. I currently exclude the individuals information if any of their parents are also in 1000G. <br>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69f0f1a0-dede-4454-a5c1-dda92d3961e8",
   "metadata": {},
   "source": [
    "### Formatting the recombination map for pipeline\n",
    "Several recombination maps exist for the human population. Each is computed using specific populations and with specific techniques.<br>\n",
    "Two major formats exist: HapMap format and PLINK format. <br>\n",
    "The main difference is how recombination rate is stored:\n",
    "- HapMap: Chromosome, Position(bp), Rate(cM/Mb)\n",
    "- PLINK: Chromosome, Cumulative Map(cM), Position\n",
    "\n",
    "Recombination rate starts at 0 at the first measured SNP position, instead of starting at the first chromosome position. <br>\n",
    "The current ~/projects/ctb-sgravel/data/genetic_maps contains many maps in Hg37 format. <br>\n",
    "I used the HapMapII map in ~/projects/ctb-sgravel/data/genetic_maps/HapMapII_GRCh38. But transformed it into cumulative Map."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c33ad72-636e-46a1-bfc1-8e2a9916c4c6",
   "metadata": {},
   "source": [
    "### Minimal example pipeline: EDAR region in chr2:107,500,000-110,000,000\n",
    "EDAR is an identified sweep in the EAS population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e3e46f3-5590-4f6c-9c9b-3b5e3248d691",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sh\n",
    "chr2_vcf_path='/home/alouette/projects/ctb-sgravel/data/1000G/20201028_CCDG_14151_B01_GRM_WGS_2020-08-05_chr2.recalibrated_variants.vcf.gz'\n",
    "strict_mask_1000G='/home/alouette/projects/ctb-sgravel/data/1000G/masks_and_annotations/20160622.allChr.mask.bed'\n",
    "IntegerT=6\n",
    "module load bcftools\n",
    "### takes 6 min\n",
    "bcftools view -i 'FILTER==\"PASS\"' -Oz -m2 -M2 -v snps -r chr2:107500000-110000000 ${chr2_vcf_path} --threads ${IntegerT} -o \"chr2.EDAR_region.vcf.gz\"\n",
    "bcftools index --threads ${IntegerT} \"chr2.EDAR_region.vcf.gz\"\n",
    "### takes 3 min\n",
    "bcftools view -Oz -R ${strict_mask_1000G} \"chr2.EDAR_region.vcf.gz\" --threads ${IntegerT} -o \"chr2.EDAR_region.strict_mask.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "48822fc5-2e63-4d9a-8f8f-ec670a0373e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf(vcf_path:str) -> dict: \n",
    "    \"\"\"\n",
    "    Read in the VCF using scikit-allel, the genotype information is stored in a dict\n",
    "    Optimization: could limit the fields read in, can also specify detailed format like float32\n",
    "    \"\"\"\n",
    "    callset = allel.read_vcf(vcf_path)\n",
    "    return callset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "723b5160-465f-4006-9709-fe6ff03f9d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_path = \"chr2.EDAR_region.strict_mask.vcf.gz\"\n",
    "EDAR_callset = read_vcf(vcf_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "39fa0af3-8ba1-4dd4-8111-def0788bc2d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'samples': array(['HG00096', 'HG00097', 'HG00099', ..., 'NA21142', 'NA21143',\n",
       "        'NA21144'], dtype=object),\n",
       " 'calldata/GT': array([[[0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]],\n",
       " \n",
       "        [[0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]],\n",
       " \n",
       "        [[0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]],\n",
       " \n",
       "        ...,\n",
       " \n",
       "        [[0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]],\n",
       " \n",
       "        [[0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]],\n",
       " \n",
       "        [[0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]]], dtype=int8),\n",
       " 'variants/ALT': array([['G', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ...,\n",
       "        ['A', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['G', '', '']], dtype=object),\n",
       " 'variants/CHROM': array(['chr2', 'chr2', 'chr2', ..., 'chr2', 'chr2', 'chr2'], dtype=object),\n",
       " 'variants/FILTER_PASS': array([ True,  True,  True, ...,  True,  True,  True]),\n",
       " 'variants/ID': array(['.', '.', '.', ..., '.', '.', '.'], dtype=object),\n",
       " 'variants/POS': array([107500027, 107500073, 107500170, ..., 109736849, 109742841,\n",
       "        109743205], dtype=int32),\n",
       " 'variants/QUAL': array([  781.65,  1768.47,   398.69, ...,   485.69, 27346.8 ,  1404.62],\n",
       "       dtype=float32),\n",
       " 'variants/REF': array(['A', 'G', 'C', ..., 'T', 'T', 'A'], dtype=object)}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EDAR_callset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "114ff70e-9e5c-44e2-9fa5-13be54995a43",
   "metadata": {},
   "outputs": [],
   "source": [
    "def msprime_read_HapMap(rec_map_path: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Require msprime >= 1.0.0\n",
    "    Read the recombination map using msprime functionality.\n",
    "    Can uses the functionality: get_cumulative_mass(), find_index()\n",
    "    Column 1-2, positions; left: inclusive, right: exclusive\n",
    "    \"\"\"\n",
    "    ms_RateMap = ms.RateMap.read_hapmap(rec_map_path, position_col=1, rate_col=2)\n",
    "    return ms_RateMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "d5ff3fa7-15ba-44f9-bd04-aef0b5e3143b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _windowing(pos_index_list: list) -> list:\n",
    "    \"\"\"\n",
    "    Master window function, based on the index list, give back a list of slice as windows that can be directly used with scikit_allel SNP_array.\n",
    "    This function is the core function and used for other detailed window_by_* functions.\n",
    "    Handles missing values, bins without any SNPs here\n",
    "    :param pos_index_list: The list of position index corresponding to the starting index of each bin\n",
    "    Return a list of slices\n",
    "    \"\"\"\n",
    "    if np.all(pos_index_list == 0):\n",
    "        raise ValueError(\"Bin position indices are all 0\")\n",
    "    window_list = []\n",
    "    for start_index, end_index in zip(pos_index_list[:-1], pos_index_list[1:]):\n",
    "        if start_index == end_index:\n",
    "            ### recombination hotspot can result in no SNPs in the slice\n",
    "            window_list.append(slice(None, -0))\n",
    "            continue\n",
    "        window_list.append(slice(start_index, end_index))\n",
    "    ### next step requires loading the data, do it when in need\n",
    "    return window_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "095a3d8b-7cdb-45f4-a844-72f8873d0a85",
   "metadata": {},
   "outputs": [],
   "source": [
    "def window_by_reombination(pos_array: list, rec_map: ms.RateMap, rec_start: float, rec_end: float, rec_step = 0.04) -> list:\n",
    "    \"\"\"\n",
    "    Require msprime >= 1.0.0 to use RateMap\n",
    "    Call this function to get the window list separated by rec_distance\n",
    "    How to handle the last position if rec subsetting is not possible?\n",
    "    \"\"\"\n",
    "    ### end has to be larger than start\n",
    "    if rec_start >= rec_end:\n",
    "        raise ValueError(\"Input recombination range is invalid\")\n",
    "    ### checking if the rec_start and rec_end are contained in the given pos_array\n",
    "    ### convert all positions to cumulative mass in cM from the rate map\n",
    "    cum_rec_mass = rec_map.get_cumulative_mass(pos_array)* 100\n",
    "    if (rec_start < np.nanmin(cum_rec_mass)) or (rec_end > np.nanmax(cum_rec_mass)):\n",
    "        raise ValueError(\"Input recombiantion range is out of range\")\n",
    "    ### if given rec distance is smaller or equal to the the step \n",
    "    ### return the original array without subsetting\n",
    "    if rec_end - rec_start <= rec_step:\n",
    "        rec_list = [rec_start, rec_end]                    \n",
    "    else:\n",
    "        rec_list = np.arange(start = rec_start, stop = rec_end, step = rec_step)\n",
    "        ### check the last position\n",
    "        ### floating point issue, the last position can be larger than end\n",
    "        if rec_list[-1] > rec_end:\n",
    "            rec_list[-1] = rec_end\n",
    "    ### get corresponding position index\n",
    "    ### currently missing values are not considered\n",
    "    pos_index_list = [np.searchsorted(cum_rec_mass, rec, side = \"left\") for rec in rec_list]\n",
    "    ### incorporate _windowing function to create slice, and also to handle missing values later\n",
    "    window_list = _windowing(pos_index_list)\n",
    "    return window_list\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "52fdff79-e306-4603-ab20-b15aed7a5f4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_array = EDAR_callset['variants/POS']\n",
    "snp_array = EDAR_callset['calldata/GT']\n",
    "rec_map_path = \"/home/alouette/projects/ctb-sgravel/data/genetic_maps/HapMapII_GRCh38/genetic_map_Hg38_chr2.txt\"\n",
    "ms_RateMap = msprime_read_HapMap(rec_map_path)\n",
    "rec_start = ms_RateMap.get_cumulative_mass(pos_array[0])*100\n",
    "rec_end = ms_RateMap.get_cumulative_mass(pos_array[-1])*100\n",
    "rec_step = 0.04\n",
    "window_list = window_by_reombination(pos_array, ms_RateMap, rec_start, rec_end, 0.04)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9270ad5f-9cf5-4671-85a8-c127503a1fef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FamilyID</th>\n",
       "      <th>SampleID</th>\n",
       "      <th>FatherID</th>\n",
       "      <th>MotherID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Population</th>\n",
       "      <th>Superpopulation</th>\n",
       "      <th>callset_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00096</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00097</td>\n",
       "      <td>HG00097</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00099</td>\n",
       "      <td>HG00099</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00100</td>\n",
       "      <td>HG00100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00101</td>\n",
       "      <td>HG00101</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3197</th>\n",
       "      <td>NA21137</td>\n",
       "      <td>NA21137</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GIH</td>\n",
       "      <td>SAS</td>\n",
       "      <td>3197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3198</th>\n",
       "      <td>NA21141</td>\n",
       "      <td>NA21141</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GIH</td>\n",
       "      <td>SAS</td>\n",
       "      <td>3198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3199</th>\n",
       "      <td>NA21142</td>\n",
       "      <td>NA21142</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GIH</td>\n",
       "      <td>SAS</td>\n",
       "      <td>3199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3200</th>\n",
       "      <td>NA21143</td>\n",
       "      <td>NA21143</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GIH</td>\n",
       "      <td>SAS</td>\n",
       "      <td>3200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3201</th>\n",
       "      <td>NA21144</td>\n",
       "      <td>NA21144</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>GIH</td>\n",
       "      <td>SAS</td>\n",
       "      <td>3201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3202 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     FamilyID SampleID FatherID MotherID  Sex Population Superpopulation  \\\n",
       "0     HG00096  HG00096        0        0    1        GBR             EUR   \n",
       "1     HG00097  HG00097        0        0    2        GBR             EUR   \n",
       "2     HG00099  HG00099        0        0    2        GBR             EUR   \n",
       "3     HG00100  HG00100        0        0    2        GBR             EUR   \n",
       "4     HG00101  HG00101        0        0    1        GBR             EUR   \n",
       "...       ...      ...      ...      ...  ...        ...             ...   \n",
       "3197  NA21137  NA21137        0        0    2        GIH             SAS   \n",
       "3198  NA21141  NA21141        0        0    2        GIH             SAS   \n",
       "3199  NA21142  NA21142        0        0    2        GIH             SAS   \n",
       "3200  NA21143  NA21143        0        0    2        GIH             SAS   \n",
       "3201  NA21144  NA21144        0        0    2        GIH             SAS   \n",
       "\n",
       "      callset_index  \n",
       "0                 0  \n",
       "1                 1  \n",
       "2                 2  \n",
       "3                 3  \n",
       "4                 4  \n",
       "...             ...  \n",
       "3197           3197  \n",
       "3198           3198  \n",
       "3199           3199  \n",
       "3200           3200  \n",
       "3201           3201  \n",
       "\n",
       "[3202 rows x 8 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "panel_path = \"/home/alouette/projects/ctb-sgravel/alouette/1000Genome/20130606_g1k_3202_samples_ped_population.txt\"\n",
    "panel = read_panel(panel_path)\n",
    "panel = add_callset_index(panel, EDAR_callset['samples'])\n",
    "panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "30c9f59c-638b-4146-a10d-240baac17f4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 48.32134167,  63.60503777, 105.63581773, 192.38169577])"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_window_H = allel.heterozygosity_observed(snp_window)\n",
    "np.sum(snp_window_LD, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "d0a2a2e1-dd4b-409a-ab3e-ff6af6b2fbe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALL_LD_list = LD_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "212feba6-96ad-4c4d-9ad0-9753612ba4ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "AFR_LD_list = []\n",
    "EUR_LD_list = []\n",
    "EAS_LD_list = []\n",
    "ALL_LD_list = []\n",
    "for window in window_list:\n",
    "    for LD_list, super_pop in zip([AFR_LD_list, EUR_LD_list, EAS_LD_list, ALL_LD_list], [\"AFR\", \"EUR\", \"EAS\", \"\"]):\n",
    "        snp_array_pop = select_by_super_population(panel, snp_array, super_pop)\n",
    "        snp_window = allel.GenotypeArray(snp_array_pop[window])\n",
    "        snp_window_012 = snp_window.to_n_alt(fill=-1)\n",
    "        snp_window_LD = moments.LD.Parsing.compute_pairwise_stats(snp_window_012, genotypes = True)\n",
    "        LD_list.append(np.sum(snp_window_LD, axis = 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "827a5d6c-53d5-480a-9b9c-d75e934cfc12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.00062461, 0.00156201, 0.0003123 , ..., 0.0003123 , 0.01681472,\n",
       "       0.00126863])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_H = allel.heterozygosity_observed(select_by_super_population(panel, snp_array, \"\"), fill = -1)\n",
    "ALL_H"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ae5f7044-9938-4b98-ad3a-150e168db517",
   "metadata": {},
   "outputs": [],
   "source": [
    "### the start and end of the window\n",
    "ALL_df = pd.DataFrame({\"pos_start\": [pos_array[window][0] for window in window_list], \n",
    "                      \"pos_end\": [pos_array[window][-1] for window in window_list],\n",
    "                      \"D2_sum\": np.array(ALL_LD_list)[:, 0],\n",
    "                      \"Dz_sum\": np.array(ALL_LD_list)[:, 1],\n",
    "                      \"Pi2_sum\": np.array(ALL_LD_list)[:, 2],\n",
    "                      \"D_sum\": np.array(ALL_LD_list)[:, 3]})\n",
    "ALL_df[\"sigD2\"] = ALL_df.D2_sum/ALL_df.Pi2_sum\n",
    "ALL_df[\"sigDz\"] = ALL_df.Dz_sum/ALL_df.Pi2_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c5d0cbba-8e6c-43ea-b20d-ffcdcc28e96c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_start</th>\n",
       "      <th>pos_end</th>\n",
       "      <th>D2_sum</th>\n",
       "      <th>Dz_sum</th>\n",
       "      <th>Pi2_sum</th>\n",
       "      <th>D_sum</th>\n",
       "      <th>sigD2</th>\n",
       "      <th>sigDz</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>107500027</td>\n",
       "      <td>107540809</td>\n",
       "      <td>4.832134e+01</td>\n",
       "      <td>63.605038</td>\n",
       "      <td>105.635818</td>\n",
       "      <td>192.381696</td>\n",
       "      <td>0.457433</td>\n",
       "      <td>0.602116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>107540817</td>\n",
       "      <td>107687099</td>\n",
       "      <td>6.354871e+02</td>\n",
       "      <td>398.262339</td>\n",
       "      <td>1803.296560</td>\n",
       "      <td>3221.582848</td>\n",
       "      <td>0.352403</td>\n",
       "      <td>0.220852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>107687103</td>\n",
       "      <td>107748695</td>\n",
       "      <td>1.938517e+02</td>\n",
       "      <td>173.704448</td>\n",
       "      <td>363.929064</td>\n",
       "      <td>1068.259842</td>\n",
       "      <td>0.532663</td>\n",
       "      <td>0.477303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>107748722</td>\n",
       "      <td>107768586</td>\n",
       "      <td>7.952209e+00</td>\n",
       "      <td>7.773271</td>\n",
       "      <td>26.283914</td>\n",
       "      <td>53.354536</td>\n",
       "      <td>0.302550</td>\n",
       "      <td>0.295743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>107768613</td>\n",
       "      <td>107811845</td>\n",
       "      <td>1.275494e+01</td>\n",
       "      <td>6.640576</td>\n",
       "      <td>67.257407</td>\n",
       "      <td>67.644414</td>\n",
       "      <td>0.189644</td>\n",
       "      <td>0.098734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>107811918</td>\n",
       "      <td>107812657</td>\n",
       "      <td>-8.097929e-09</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>-0.000056</td>\n",
       "      <td>0.140119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>107812721</td>\n",
       "      <td>107815597</td>\n",
       "      <td>5.994969e-02</td>\n",
       "      <td>0.114484</td>\n",
       "      <td>0.355099</td>\n",
       "      <td>0.895032</td>\n",
       "      <td>0.168825</td>\n",
       "      <td>0.322399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>107815617</td>\n",
       "      <td>107995173</td>\n",
       "      <td>5.849300e+01</td>\n",
       "      <td>61.258251</td>\n",
       "      <td>253.778178</td>\n",
       "      <td>377.780461</td>\n",
       "      <td>0.230489</td>\n",
       "      <td>0.241385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>107995201</td>\n",
       "      <td>108010232</td>\n",
       "      <td>2.869461e-01</td>\n",
       "      <td>0.745619</td>\n",
       "      <td>2.949769</td>\n",
       "      <td>1.817141</td>\n",
       "      <td>0.097278</td>\n",
       "      <td>0.252772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>108010302</td>\n",
       "      <td>108028349</td>\n",
       "      <td>2.368929e+00</td>\n",
       "      <td>2.196657</td>\n",
       "      <td>9.165350</td>\n",
       "      <td>1.272175</td>\n",
       "      <td>0.258466</td>\n",
       "      <td>0.239670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>108028409</td>\n",
       "      <td>108103631</td>\n",
       "      <td>9.167368e+01</td>\n",
       "      <td>113.563724</td>\n",
       "      <td>348.332132</td>\n",
       "      <td>301.678840</td>\n",
       "      <td>0.263179</td>\n",
       "      <td>0.326021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>108103680</td>\n",
       "      <td>108106839</td>\n",
       "      <td>4.473245e-02</td>\n",
       "      <td>0.162106</td>\n",
       "      <td>0.384989</td>\n",
       "      <td>0.079649</td>\n",
       "      <td>0.116191</td>\n",
       "      <td>0.421067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>108106900</td>\n",
       "      <td>108164218</td>\n",
       "      <td>8.343960e+01</td>\n",
       "      <td>48.258651</td>\n",
       "      <td>236.418576</td>\n",
       "      <td>73.844527</td>\n",
       "      <td>0.352932</td>\n",
       "      <td>0.204124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>108164249</td>\n",
       "      <td>108278513</td>\n",
       "      <td>4.047532e+02</td>\n",
       "      <td>110.285279</td>\n",
       "      <td>1288.043088</td>\n",
       "      <td>1212.690775</td>\n",
       "      <td>0.314239</td>\n",
       "      <td>0.085622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>108278542</td>\n",
       "      <td>108311342</td>\n",
       "      <td>5.447462e+00</td>\n",
       "      <td>15.259440</td>\n",
       "      <td>40.759675</td>\n",
       "      <td>30.748407</td>\n",
       "      <td>0.133648</td>\n",
       "      <td>0.374376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>108311512</td>\n",
       "      <td>108312757</td>\n",
       "      <td>3.964425e-04</td>\n",
       "      <td>0.013516</td>\n",
       "      <td>0.005200</td>\n",
       "      <td>0.015761</td>\n",
       "      <td>0.076244</td>\n",
       "      <td>2.599364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>108312790</td>\n",
       "      <td>108314070</td>\n",
       "      <td>3.726702e-03</td>\n",
       "      <td>0.084293</td>\n",
       "      <td>0.052922</td>\n",
       "      <td>0.132403</td>\n",
       "      <td>0.070419</td>\n",
       "      <td>1.592769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>108314216</td>\n",
       "      <td>108320948</td>\n",
       "      <td>4.113654e-01</td>\n",
       "      <td>1.363112</td>\n",
       "      <td>2.021434</td>\n",
       "      <td>3.063529</td>\n",
       "      <td>0.203502</td>\n",
       "      <td>0.674329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>108321084</td>\n",
       "      <td>108331369</td>\n",
       "      <td>7.723038e+00</td>\n",
       "      <td>5.044640</td>\n",
       "      <td>26.473238</td>\n",
       "      <td>36.939942</td>\n",
       "      <td>0.291730</td>\n",
       "      <td>0.190556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>108331393</td>\n",
       "      <td>108333066</td>\n",
       "      <td>3.432664e-05</td>\n",
       "      <td>0.008460</td>\n",
       "      <td>0.005256</td>\n",
       "      <td>0.009238</td>\n",
       "      <td>0.006531</td>\n",
       "      <td>1.609483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>108333128</td>\n",
       "      <td>108473334</td>\n",
       "      <td>3.082764e+02</td>\n",
       "      <td>396.928961</td>\n",
       "      <td>967.413871</td>\n",
       "      <td>922.089112</td>\n",
       "      <td>0.318660</td>\n",
       "      <td>0.410299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>108473428</td>\n",
       "      <td>108719736</td>\n",
       "      <td>5.911510e+02</td>\n",
       "      <td>98.323506</td>\n",
       "      <td>1113.031308</td>\n",
       "      <td>1305.293597</td>\n",
       "      <td>0.531118</td>\n",
       "      <td>0.088338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>108721083</td>\n",
       "      <td>108923676</td>\n",
       "      <td>2.594268e+02</td>\n",
       "      <td>727.249573</td>\n",
       "      <td>808.014188</td>\n",
       "      <td>297.032072</td>\n",
       "      <td>0.321067</td>\n",
       "      <td>0.900046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>108923775</td>\n",
       "      <td>108925012</td>\n",
       "      <td>1.774266e-04</td>\n",
       "      <td>0.010327</td>\n",
       "      <td>0.003167</td>\n",
       "      <td>0.011556</td>\n",
       "      <td>0.056017</td>\n",
       "      <td>3.260513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>108925038</td>\n",
       "      <td>108926747</td>\n",
       "      <td>1.073408e-02</td>\n",
       "      <td>0.142756</td>\n",
       "      <td>0.168948</td>\n",
       "      <td>0.199006</td>\n",
       "      <td>0.063535</td>\n",
       "      <td>0.844968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>108926778</td>\n",
       "      <td>108966930</td>\n",
       "      <td>5.617797e+01</td>\n",
       "      <td>23.256495</td>\n",
       "      <td>76.419064</td>\n",
       "      <td>38.574167</td>\n",
       "      <td>0.735130</td>\n",
       "      <td>0.304328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>108967448</td>\n",
       "      <td>108988354</td>\n",
       "      <td>1.479267e+01</td>\n",
       "      <td>5.041021</td>\n",
       "      <td>43.978008</td>\n",
       "      <td>24.836396</td>\n",
       "      <td>0.336365</td>\n",
       "      <td>0.114626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>108988410</td>\n",
       "      <td>109011154</td>\n",
       "      <td>6.538866e+00</td>\n",
       "      <td>1.373934</td>\n",
       "      <td>20.385880</td>\n",
       "      <td>9.491561</td>\n",
       "      <td>0.320755</td>\n",
       "      <td>0.067396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>109011262</td>\n",
       "      <td>109012378</td>\n",
       "      <td>3.544129e-04</td>\n",
       "      <td>0.016966</td>\n",
       "      <td>0.002519</td>\n",
       "      <td>0.018705</td>\n",
       "      <td>0.140674</td>\n",
       "      <td>6.734056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>109012381</td>\n",
       "      <td>109023176</td>\n",
       "      <td>1.234226e+00</td>\n",
       "      <td>3.855376</td>\n",
       "      <td>4.857996</td>\n",
       "      <td>8.388759</td>\n",
       "      <td>0.254061</td>\n",
       "      <td>0.793615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>109023212</td>\n",
       "      <td>109038957</td>\n",
       "      <td>3.950941e-01</td>\n",
       "      <td>3.785456</td>\n",
       "      <td>5.550180</td>\n",
       "      <td>2.170360</td>\n",
       "      <td>0.071186</td>\n",
       "      <td>0.682042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>109038992</td>\n",
       "      <td>109060069</td>\n",
       "      <td>2.474614e+00</td>\n",
       "      <td>11.208526</td>\n",
       "      <td>11.014197</td>\n",
       "      <td>7.103188</td>\n",
       "      <td>0.224675</td>\n",
       "      <td>1.017644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>109060084</td>\n",
       "      <td>109086154</td>\n",
       "      <td>1.009840e+01</td>\n",
       "      <td>22.333559</td>\n",
       "      <td>50.496342</td>\n",
       "      <td>33.837493</td>\n",
       "      <td>0.199983</td>\n",
       "      <td>0.442281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>109086211</td>\n",
       "      <td>109204465</td>\n",
       "      <td>1.285981e+02</td>\n",
       "      <td>161.368764</td>\n",
       "      <td>677.192339</td>\n",
       "      <td>403.377091</td>\n",
       "      <td>0.189899</td>\n",
       "      <td>0.238291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>109204534</td>\n",
       "      <td>109300640</td>\n",
       "      <td>9.805824e+01</td>\n",
       "      <td>84.283276</td>\n",
       "      <td>578.349985</td>\n",
       "      <td>488.333826</td>\n",
       "      <td>0.169548</td>\n",
       "      <td>0.145731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>109300688</td>\n",
       "      <td>109343434</td>\n",
       "      <td>4.689265e+02</td>\n",
       "      <td>79.960197</td>\n",
       "      <td>766.823149</td>\n",
       "      <td>1367.866059</td>\n",
       "      <td>0.611518</td>\n",
       "      <td>0.104275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>109343443</td>\n",
       "      <td>109385748</td>\n",
       "      <td>2.879279e+01</td>\n",
       "      <td>96.920691</td>\n",
       "      <td>173.060152</td>\n",
       "      <td>213.962576</td>\n",
       "      <td>0.166374</td>\n",
       "      <td>0.560040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>109385847</td>\n",
       "      <td>109386969</td>\n",
       "      <td>1.694228e-02</td>\n",
       "      <td>-0.005996</td>\n",
       "      <td>0.114208</td>\n",
       "      <td>0.149365</td>\n",
       "      <td>0.148346</td>\n",
       "      <td>-0.052504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>109387003</td>\n",
       "      <td>109433696</td>\n",
       "      <td>8.292654e+01</td>\n",
       "      <td>38.830993</td>\n",
       "      <td>392.652963</td>\n",
       "      <td>336.469071</td>\n",
       "      <td>0.211196</td>\n",
       "      <td>0.098894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>109433724</td>\n",
       "      <td>109484026</td>\n",
       "      <td>2.018424e+01</td>\n",
       "      <td>50.741941</td>\n",
       "      <td>85.803407</td>\n",
       "      <td>118.555144</td>\n",
       "      <td>0.235238</td>\n",
       "      <td>0.591374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>109484401</td>\n",
       "      <td>109490793</td>\n",
       "      <td>2.196724e-01</td>\n",
       "      <td>0.436423</td>\n",
       "      <td>1.649977</td>\n",
       "      <td>0.082108</td>\n",
       "      <td>0.133137</td>\n",
       "      <td>0.264502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>109490849</td>\n",
       "      <td>109497715</td>\n",
       "      <td>2.293102e+01</td>\n",
       "      <td>15.489118</td>\n",
       "      <td>33.748275</td>\n",
       "      <td>70.597041</td>\n",
       "      <td>0.679473</td>\n",
       "      <td>0.458960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>109497769</td>\n",
       "      <td>109501728</td>\n",
       "      <td>4.120900e+00</td>\n",
       "      <td>2.752371</td>\n",
       "      <td>13.701297</td>\n",
       "      <td>20.833319</td>\n",
       "      <td>0.300767</td>\n",
       "      <td>0.200884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>109501752</td>\n",
       "      <td>109503211</td>\n",
       "      <td>1.232959e-04</td>\n",
       "      <td>0.002091</td>\n",
       "      <td>0.022829</td>\n",
       "      <td>0.001098</td>\n",
       "      <td>0.005401</td>\n",
       "      <td>0.091593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>109503288</td>\n",
       "      <td>109504830</td>\n",
       "      <td>1.600396e-02</td>\n",
       "      <td>0.048415</td>\n",
       "      <td>0.119330</td>\n",
       "      <td>0.220826</td>\n",
       "      <td>0.134116</td>\n",
       "      <td>0.405721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>109504907</td>\n",
       "      <td>109515614</td>\n",
       "      <td>2.446952e+00</td>\n",
       "      <td>1.194289</td>\n",
       "      <td>15.327350</td>\n",
       "      <td>1.910507</td>\n",
       "      <td>0.159646</td>\n",
       "      <td>0.077919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>109515696</td>\n",
       "      <td>109653711</td>\n",
       "      <td>1.626474e+02</td>\n",
       "      <td>204.819008</td>\n",
       "      <td>976.489941</td>\n",
       "      <td>500.195523</td>\n",
       "      <td>0.166563</td>\n",
       "      <td>0.209750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>109653838</td>\n",
       "      <td>109678414</td>\n",
       "      <td>1.419048e+00</td>\n",
       "      <td>0.641122</td>\n",
       "      <td>13.583871</td>\n",
       "      <td>5.533742</td>\n",
       "      <td>0.104466</td>\n",
       "      <td>0.047197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>109678446</td>\n",
       "      <td>109682579</td>\n",
       "      <td>1.148446e-01</td>\n",
       "      <td>0.153616</td>\n",
       "      <td>1.037618</td>\n",
       "      <td>0.214775</td>\n",
       "      <td>0.110681</td>\n",
       "      <td>0.148047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>109682890</td>\n",
       "      <td>109687723</td>\n",
       "      <td>5.243255e-02</td>\n",
       "      <td>-0.002537</td>\n",
       "      <td>0.383900</td>\n",
       "      <td>0.034232</td>\n",
       "      <td>0.136579</td>\n",
       "      <td>-0.006608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>109688220</td>\n",
       "      <td>109691762</td>\n",
       "      <td>1.724442e-01</td>\n",
       "      <td>-0.044805</td>\n",
       "      <td>1.402038</td>\n",
       "      <td>-0.160746</td>\n",
       "      <td>0.122995</td>\n",
       "      <td>-0.031957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>109692129</td>\n",
       "      <td>109694787</td>\n",
       "      <td>1.846644e-02</td>\n",
       "      <td>0.231545</td>\n",
       "      <td>0.318992</td>\n",
       "      <td>0.159465</td>\n",
       "      <td>0.057890</td>\n",
       "      <td>0.725864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>109694890</td>\n",
       "      <td>109703793</td>\n",
       "      <td>1.477016e+00</td>\n",
       "      <td>1.265391</td>\n",
       "      <td>3.997957</td>\n",
       "      <td>2.448511</td>\n",
       "      <td>0.369443</td>\n",
       "      <td>0.316509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>109703850</td>\n",
       "      <td>109718428</td>\n",
       "      <td>1.652565e+00</td>\n",
       "      <td>1.129501</td>\n",
       "      <td>4.550031</td>\n",
       "      <td>6.471406</td>\n",
       "      <td>0.363199</td>\n",
       "      <td>0.248240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>109718473</td>\n",
       "      <td>109733043</td>\n",
       "      <td>2.720191e-01</td>\n",
       "      <td>0.378740</td>\n",
       "      <td>0.955010</td>\n",
       "      <td>0.647343</td>\n",
       "      <td>0.284834</td>\n",
       "      <td>0.396582</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pos_start    pos_end        D2_sum      Dz_sum      Pi2_sum        D_sum  \\\n",
       "0   107500027  107540809  4.832134e+01   63.605038   105.635818   192.381696   \n",
       "1   107540817  107687099  6.354871e+02  398.262339  1803.296560  3221.582848   \n",
       "2   107687103  107748695  1.938517e+02  173.704448   363.929064  1068.259842   \n",
       "3   107748722  107768586  7.952209e+00    7.773271    26.283914    53.354536   \n",
       "4   107768613  107811845  1.275494e+01    6.640576    67.257407    67.644414   \n",
       "5   107811918  107812657 -8.097929e-09    0.000020     0.000145     0.000019   \n",
       "6   107812721  107815597  5.994969e-02    0.114484     0.355099     0.895032   \n",
       "7   107815617  107995173  5.849300e+01   61.258251   253.778178   377.780461   \n",
       "8   107995201  108010232  2.869461e-01    0.745619     2.949769     1.817141   \n",
       "9   108010302  108028349  2.368929e+00    2.196657     9.165350     1.272175   \n",
       "10  108028409  108103631  9.167368e+01  113.563724   348.332132   301.678840   \n",
       "11  108103680  108106839  4.473245e-02    0.162106     0.384989     0.079649   \n",
       "12  108106900  108164218  8.343960e+01   48.258651   236.418576    73.844527   \n",
       "13  108164249  108278513  4.047532e+02  110.285279  1288.043088  1212.690775   \n",
       "14  108278542  108311342  5.447462e+00   15.259440    40.759675    30.748407   \n",
       "15  108311512  108312757  3.964425e-04    0.013516     0.005200     0.015761   \n",
       "16  108312790  108314070  3.726702e-03    0.084293     0.052922     0.132403   \n",
       "17  108314216  108320948  4.113654e-01    1.363112     2.021434     3.063529   \n",
       "18  108321084  108331369  7.723038e+00    5.044640    26.473238    36.939942   \n",
       "19  108331393  108333066  3.432664e-05    0.008460     0.005256     0.009238   \n",
       "20  108333128  108473334  3.082764e+02  396.928961   967.413871   922.089112   \n",
       "21  108473428  108719736  5.911510e+02   98.323506  1113.031308  1305.293597   \n",
       "22  108721083  108923676  2.594268e+02  727.249573   808.014188   297.032072   \n",
       "23  108923775  108925012  1.774266e-04    0.010327     0.003167     0.011556   \n",
       "24  108925038  108926747  1.073408e-02    0.142756     0.168948     0.199006   \n",
       "25  108926778  108966930  5.617797e+01   23.256495    76.419064    38.574167   \n",
       "26  108967448  108988354  1.479267e+01    5.041021    43.978008    24.836396   \n",
       "27  108988410  109011154  6.538866e+00    1.373934    20.385880     9.491561   \n",
       "28  109011262  109012378  3.544129e-04    0.016966     0.002519     0.018705   \n",
       "29  109012381  109023176  1.234226e+00    3.855376     4.857996     8.388759   \n",
       "30  109023212  109038957  3.950941e-01    3.785456     5.550180     2.170360   \n",
       "31  109038992  109060069  2.474614e+00   11.208526    11.014197     7.103188   \n",
       "32  109060084  109086154  1.009840e+01   22.333559    50.496342    33.837493   \n",
       "33  109086211  109204465  1.285981e+02  161.368764   677.192339   403.377091   \n",
       "34  109204534  109300640  9.805824e+01   84.283276   578.349985   488.333826   \n",
       "35  109300688  109343434  4.689265e+02   79.960197   766.823149  1367.866059   \n",
       "36  109343443  109385748  2.879279e+01   96.920691   173.060152   213.962576   \n",
       "37  109385847  109386969  1.694228e-02   -0.005996     0.114208     0.149365   \n",
       "38  109387003  109433696  8.292654e+01   38.830993   392.652963   336.469071   \n",
       "39  109433724  109484026  2.018424e+01   50.741941    85.803407   118.555144   \n",
       "40  109484401  109490793  2.196724e-01    0.436423     1.649977     0.082108   \n",
       "41  109490849  109497715  2.293102e+01   15.489118    33.748275    70.597041   \n",
       "42  109497769  109501728  4.120900e+00    2.752371    13.701297    20.833319   \n",
       "43  109501752  109503211  1.232959e-04    0.002091     0.022829     0.001098   \n",
       "44  109503288  109504830  1.600396e-02    0.048415     0.119330     0.220826   \n",
       "45  109504907  109515614  2.446952e+00    1.194289    15.327350     1.910507   \n",
       "46  109515696  109653711  1.626474e+02  204.819008   976.489941   500.195523   \n",
       "47  109653838  109678414  1.419048e+00    0.641122    13.583871     5.533742   \n",
       "48  109678446  109682579  1.148446e-01    0.153616     1.037618     0.214775   \n",
       "49  109682890  109687723  5.243255e-02   -0.002537     0.383900     0.034232   \n",
       "50  109688220  109691762  1.724442e-01   -0.044805     1.402038    -0.160746   \n",
       "51  109692129  109694787  1.846644e-02    0.231545     0.318992     0.159465   \n",
       "52  109694890  109703793  1.477016e+00    1.265391     3.997957     2.448511   \n",
       "53  109703850  109718428  1.652565e+00    1.129501     4.550031     6.471406   \n",
       "54  109718473  109733043  2.720191e-01    0.378740     0.955010     0.647343   \n",
       "\n",
       "       sigD2     sigDz  \n",
       "0   0.457433  0.602116  \n",
       "1   0.352403  0.220852  \n",
       "2   0.532663  0.477303  \n",
       "3   0.302550  0.295743  \n",
       "4   0.189644  0.098734  \n",
       "5  -0.000056  0.140119  \n",
       "6   0.168825  0.322399  \n",
       "7   0.230489  0.241385  \n",
       "8   0.097278  0.252772  \n",
       "9   0.258466  0.239670  \n",
       "10  0.263179  0.326021  \n",
       "11  0.116191  0.421067  \n",
       "12  0.352932  0.204124  \n",
       "13  0.314239  0.085622  \n",
       "14  0.133648  0.374376  \n",
       "15  0.076244  2.599364  \n",
       "16  0.070419  1.592769  \n",
       "17  0.203502  0.674329  \n",
       "18  0.291730  0.190556  \n",
       "19  0.006531  1.609483  \n",
       "20  0.318660  0.410299  \n",
       "21  0.531118  0.088338  \n",
       "22  0.321067  0.900046  \n",
       "23  0.056017  3.260513  \n",
       "24  0.063535  0.844968  \n",
       "25  0.735130  0.304328  \n",
       "26  0.336365  0.114626  \n",
       "27  0.320755  0.067396  \n",
       "28  0.140674  6.734056  \n",
       "29  0.254061  0.793615  \n",
       "30  0.071186  0.682042  \n",
       "31  0.224675  1.017644  \n",
       "32  0.199983  0.442281  \n",
       "33  0.189899  0.238291  \n",
       "34  0.169548  0.145731  \n",
       "35  0.611518  0.104275  \n",
       "36  0.166374  0.560040  \n",
       "37  0.148346 -0.052504  \n",
       "38  0.211196  0.098894  \n",
       "39  0.235238  0.591374  \n",
       "40  0.133137  0.264502  \n",
       "41  0.679473  0.458960  \n",
       "42  0.300767  0.200884  \n",
       "43  0.005401  0.091593  \n",
       "44  0.134116  0.405721  \n",
       "45  0.159646  0.077919  \n",
       "46  0.166563  0.209750  \n",
       "47  0.104466  0.047197  \n",
       "48  0.110681  0.148047  \n",
       "49  0.136579 -0.006608  \n",
       "50  0.122995 -0.031957  \n",
       "51  0.057890  0.725864  \n",
       "52  0.369443  0.316509  \n",
       "53  0.363199  0.248240  \n",
       "54  0.284834  0.396582  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "946af9d7-cfdc-4a3e-8694-bf162789a9c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_callset_index(panel: pd.DataFrame, callset_sample: list):\n",
    "    samples_list = list(callset_sample)\n",
    "    samples_callset_index = [samples_list.index(s) for s in panel.SampleID]\n",
    "    panel['callset_index'] = samples_callset_index\n",
    "    return panel\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "9504e3c9-f06e-4b67-aa5a-8933b63fd66a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_panel(panel_path:str) -> pd.DataFrame:\n",
    "    panel = pd.read_csv(panel_path, sep=' ')\n",
    "    return panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "589814c2-1a04-41f2-9279-66a6e1240405",
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_by_super_population(panel: pd.DataFrame, snp_array: list, super_population :str) -> list:\n",
    "    \"\"\"\n",
    "    This will not affect the snp index position\n",
    "    \"\"\"\n",
    "    if len(super_population) == 0:\n",
    "        return allel.GenotypeArray(snp_array)\n",
    "    loc_sample = panel[panel.Superpopulation == super_population].callset_index.values\n",
    "    return snp_array.take(loc_sample, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "faae4554-6e82-40cb-ae44-2f44aa77c8c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAs2D\"><span>&lt;GenotypeArray shape=(58860, 633, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">0</th><th style=\"text-align: center\">1</th><th style=\"text-align: center\">2</th><th style=\"text-align: center\">3</th><th style=\"text-align: center\">4</th><th style=\"text-align: center\">...</th><th style=\"text-align: center\">628</th><th style=\"text-align: center\">629</th><th style=\"text-align: center\">630</th><th style=\"text-align: center\">631</th><th style=\"text-align: center\">632</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"12\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">58857</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">58858</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">58859</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<GenotypeArray shape=(58860, 633, 2) dtype=int8>\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "...\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "select_by_super_population(panel, snp_array, \"EUR\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Dz_moments_kernel",
   "language": "python",
   "name": "dz_moments"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
